From ff887060874fe50bf9b2ca0778076961390613d5 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Tue, 31 Jul 2018 21:39:27 +0800
Subject: [PATCH 4293/5242] MGS-3255-43 [#ccc] fix gpu user memory corruption

commit  47c734c108e37f9c74b182300bb40be4a356ac2d from
https://source.codeaurora.org/external/imx/linux-imx.git

fix memory overwrite since out of range in last page,
remove dma_sync_single_for_device to avoid over-touch.

this fix QM gpu crash with test_buffers stress case.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../default/gc_hal_kernel_allocator_user_memory.c  |   21 ++++++--------------
 1 file changed, 6 insertions(+), 15 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
index 301db84..578af3f 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
@@ -201,21 +201,6 @@ static int import_page_map(struct um_desc *um,
         goto error;
     }
 
-    if (addr & ~PAGE_MASK)
-    {
-        dma_sync_single_for_device(galcore_device,
-                                   page_to_phys(pages[0]),
-                                   PAGE_SIZE,
-                                   DMA_TO_DEVICE);
-    }
-    if (page_count > 1 && ((addr + size) & ~PAGE_MASK))
-    {
-        dma_sync_single_for_device(galcore_device,
-                                   page_to_phys(pages[page_count-1]),
-                                   PAGE_SIZE,
-                                   DMA_TO_DEVICE);
-    }
-
     um->type = UM_PAGE_MAP;
     um->pages = pages;
 
@@ -405,6 +390,12 @@ static int import_pfn_map(struct um_desc *um,
             get_user(data, (u32 *)vaddr);
             put_user(data, (u32 *)vaddr);
             vaddr += PAGE_SIZE;
+
+            /* Fix QM crash with test_buffers */
+            if (vaddr > memory + Size - 4)
+            {
+                vaddr = memory + Size - 4;
+            }
         }
 
         vma = find_vma(current->mm, memory);
-- 
1.7.9.5

